### PR TITLE
task, percpu: Use VMR to manage the task pagetable and VMKernelStack to allocate the task kernel stack

### PR DESCRIPTION
**Pull request by roy-hopkins**
Now that virtual memory ranges and page table entries can be managed using the new `VMR` and associated structs, update the Task structure implementation to manage the task pagetable using these new capabilities.

The task structure now includes a `VMR` instance that defines the virtual address range for the task. The kernel stack for the task is created using `VMKernelStack` and added to the task VMR. This is applied to the task pagetable during creation.

`PerCpu` already includes a per cpu VMR. This needs to be applied to the task when it is context switched onto the CPU. Currently this is done directly by copying the PerCpu level 3 index. This patch adss the function `populate_page_table()` to PerCpu which applies the PerCpu VMR to the pagetable instead.

### CODE REVIEW

**Comment by roy-hopkins** [comment]
You can see this being used in #64

File: src/task/tasks.rs [hunk 1]
```diff
@@ -183,12 +170,12 @@ pub struct TaskContext {
...
     pub rsp: u64,
 
-    /// Information about the task stack
-    pub stack: TaskStack,
-
     /// Page table that is loaded when the task is scheduled
     pub page_table: SpinLock<PageTableRef>,
 
+    /// Task virtual memory range
+    vm_range: VMR,
```
Comment (joergroedel) [inline src/task/tasks.rs:177]: This needs a more specific name, as it covers the kernel VMR. We will get another VMR for user-space in the same structure. Maybe `vm_kernel_range`?

File: src/cpu/percpu.rs [hunk 2]
```diff
@@ -171,6 +171,29 @@ impl GuestVmsaRef {
...
+}
+
+impl Drop for PerCpuMapping {
+    fn drop(&mut self) {
+        this_cpu_mut()
+            .vm_range
+            .remove(self.0)
+            .expect("Failed to remove PerCpuMapping managed range");
+    }
+}
```
Comment (joergroedel) [inline src/cpu/percpu.rs:195]: Maybe we can look into making this more generic to work with any VMR instance. This requires taking a reference to the VMR within the struct and lifetime annotations. But this is just a note, no requirement for this PR.

**Review by joergroedel — CHANGES_REQUESTED** [review]
Looks good overall, thanks. I requested one change and made one suggestion for the mapping guard (but it is fine as-is for now).

File: src/task/tasks.rs (hunk 1 above)
Comment (roy-hopkins) [reply src/task/tasks.rs]: Renamed as suggested.

File: src/cpu/percpu.rs (hunk 2 above)
Comment (roy-hopkins) [reply src/cpu/percpu.rs]: I didn't really like that struct sitting there in percpu.rs so as suggested I've added the generic function using a couple of new commits. Please take a look.

**Review by joergroedel — APPROVED** [review]
Looks good now, thanks!

**Pull request merged**